runtime/notification.go

```diff
@@ -0,0 +1,71 @@
+package runtime
+
+import (
+	"fmt"
+
+	"github.com/mitchellh/mapstructure"
+)
+
+type Notification struct {
```

Comment:
I have doubts about calling this `Notification`. That term is very much used in UX/user-facing contexts to refer to methods that push information to the user that they are not expecting. I would just call this `Event`.

Reply:
In fact, this whole type seems very redundant to the existing runtime.Event type, which is also json serializable. I don't see a reason why we can't just add `RecordStageStart()/RecordStageEnd(Outcome)` to the `RunEnv`, and its corresponding `EventTypes` instead of creating this other data model.